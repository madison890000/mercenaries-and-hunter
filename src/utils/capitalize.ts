const capitalize = (chars?: string) => {
    if (!chars) {
        return '';
    }
    const [first, ...rest] = chars.split('');
    return first.toUpperCase() + rest.join('');
};

export default capitalize;
