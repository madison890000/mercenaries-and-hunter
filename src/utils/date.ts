import {MONTH_NUMBER_OF_ONE_YEAR, READABLE_MONTHS} from '../constants/date';
import {Dayjs} from "dayjs";

const translateToReadableMonth = (month: number) => READABLE_MONTHS[month]?.slice(0, 3);

export const formatDate = (time: Dayjs) => {
    return `${translateToReadableMonth(time.month())}, ${time.year()}`;
};

const getFullMonth = (time: Dayjs) => {
    return time.year() * MONTH_NUMBER_OF_ONE_YEAR + time.month();
};

export function findPeriodByDate<T>(time: Dayjs, periods: (T & { start: Dayjs; end?: Dayjs })[]) {
    return periods?.find(p => {
        const notBefore = getFullMonth(p?.start) <= getFullMonth(time);
        const notAfter = getFullMonth(p?.end ?? new Dayjs()) >= getFullMonth(time);
        return notBefore && notAfter;
    });
}

export const getMonthCountFromStartAndEnd = (start: Dayjs, end: Dayjs) => {
    const years = end.year() - start.year();
    if (years) {
        return MONTH_NUMBER_OF_ONE_YEAR - start.month() + (years - 1) * MONTH_NUMBER_OF_ONE_YEAR + end.month();
    }
    return end.month() - start.month();
};
