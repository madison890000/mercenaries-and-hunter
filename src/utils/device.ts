export const isMobilePhone = () => window.innerWidth < 500
