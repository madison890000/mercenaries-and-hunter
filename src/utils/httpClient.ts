import axios, {AxiosError} from "axios";
import {API_DOMAIN} from "../constants/domain";
import {getGoogleToken} from "./index";

const httpClient = axios.create({
    baseURL: API_DOMAIN,
})

const handleResponse = function (response: any) {
    return response;
}

const handleError = function (error: AxiosError) {
    return Promise.reject(error);
}
httpClient.interceptors.request.use((req) => {
    const token = getGoogleToken();
    if (token) {
        req.headers.set('authorization', `Bearer ${token}`);
    }
    return req
})
httpClient.interceptors.response.use(handleResponse, handleError);

export default httpClient