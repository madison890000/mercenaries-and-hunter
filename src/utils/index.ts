
export const saveItem = (key: string, value: string) => {
  window.localStorage.setItem(key, value);
};

export const getItem = (key: string, type = 'string') => {
  return new Promise(resolve => {
    const data = window.localStorage.getItem(key);
    resolve(data);
  });
};
