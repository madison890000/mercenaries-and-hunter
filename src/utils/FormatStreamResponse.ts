import { TIME_OUT_MS } from '../constants/date';

const formatStreamResponse = async (
  res: Response,
  {
    onProgress,
    onFinish,
    onError
  }: {
    onProgress?: (e: string) => void;
    onFinish?: (e: string) => void;
    onError?: (e: Error) => void;
  }
) => {
  let responseText = '';
  onProgress?.(responseText);
  if (res.ok) {
    const reader = res.body?.getReader();
    const decoder = new TextDecoder();
    while (true) {
      const resTimeoutId = setTimeout(() => onFinish?.(''), TIME_OUT_MS);
      const content = await reader?.read();
      clearTimeout(resTimeoutId);
      if (!content || !content.value) {
        break;
      }
      const text = decoder.decode(content.value, { stream: true });
      responseText += text;
      const done = content.done;
      onProgress?.(responseText);
      if (done) {
        break;
      }
    }

    onFinish?.(responseText);
  } else if (res.status === 401) {
    console.error('Unauthorized');
    onError?.(new Error('Unauthorized'));
  } else {
    onError?.(new Error('Stream Error'));
  }
};

export default formatStreamResponse;
