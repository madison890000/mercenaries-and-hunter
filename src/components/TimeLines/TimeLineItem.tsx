import React, { useMemo } from 'react';
import { getMonthCountFromStartAndEnd } from '../../utils/date';
import MonthRect from './MonthRect';
import { getScreenDevice, ScreenDevice } from '../../utils/device';
import { BASE_MONTH_SVG_HEIGHT, BASE_RECT_WIDTH, RECT_GAP } from '../../constants/widths';

export const getBaseRectFull = (device: ScreenDevice) => {
    switch (device) {
        case ScreenDevice.A4:
            return BASE_RECT_WIDTH - 2;
        case ScreenDevice.Mobile:
            return BASE_RECT_WIDTH - 4;
        case ScreenDevice.PC:
            return BASE_RECT_WIDTH;
        default:
            return BASE_RECT_WIDTH;
    }
};

const TimeLineItem = ({ start, end, periodColor }: { start: Date; end?: Date; periodColor: string }) => {
    const { rectFullWidth, totalRects, rects, rectWidth } = useMemo(() => {
        const totalRects = getMonthCountFromStartAndEnd(start, end ?? new Date()) + 1;
        const rectFullWidth = getBaseRectFull(getScreenDevice());
        const rectWidth = rectFullWidth - RECT_GAP;
        let rects = new Array(totalRects).fill(0);
        return {
            rectFullWidth,
            totalRects,
            rects,
            rectWidth
        };
    }, [start, end]);
    return (
        <svg width={rectFullWidth * totalRects} height={BASE_MONTH_SVG_HEIGHT}>
            <g>
                {rects?.map((e, index) => (
                    <MonthRect key={index} width={rectWidth} gap={RECT_GAP} color={periodColor} index={index} />
                ))}
            </g>
        </svg>
    );
};

export default TimeLineItem;
