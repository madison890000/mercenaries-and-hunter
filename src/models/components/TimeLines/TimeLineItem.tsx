import React, {useMemo} from 'react';
import {getMonthCountFromStartAndEnd} from '../../../utils/date';
import MonthRect from './MonthRect';
import {ScreenDevice} from '../../../utils/device';
import {BASE_MONTH_SVG_HEIGHT, BASE_RECT_WIDTH, RECT_GAP} from '../../../constants/widths';
import {Dayjs} from "dayjs";

export const getBaseRectFull = (device: ScreenDevice) => {
    switch (device) {
        case ScreenDevice.A4:
            return BASE_RECT_WIDTH - 2;
        case ScreenDevice.Mobile:
            return BASE_RECT_WIDTH - 4;
        case ScreenDevice.PC:
            return BASE_RECT_WIDTH;
        default:
            return BASE_RECT_WIDTH;
    }
};

const TimeLineItem = ({start, end, periodColor}: { start: Dayjs; end?: Dayjs; periodColor: string }) => {
    const {totalRects, rects} = useMemo(() => {
        const totalRects = getMonthCountFromStartAndEnd(start, end ?? new Dayjs()) + 1;
        let rects = new Array(totalRects).fill(0);
        return {
            totalRects,
            rects
        };
    }, [start, end]);
    return (
        <svg width={BASE_MONTH_SVG_HEIGHT} height={(10 + 1) * totalRects}>
            <g>
                {rects?.map((e, index) => (
                    <MonthRect variant="v" key={index} width={10} gap={RECT_GAP} color={periodColor} index={index}/>
                ))}
            </g>
        </svg>
    );
};

export default TimeLineItem;
