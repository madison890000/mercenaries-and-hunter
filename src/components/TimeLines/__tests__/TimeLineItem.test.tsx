import React from 'react';
import TimeLineItem,{getBaseRectFull} from '../TimeLineItem';
import renderer from 'react-test-renderer';
import { ScreenDevice } from '../../../utils/device';

test('renders TimeLineItem with no end', () => {
    const TimeLineItemDom = renderer.create(
        <TimeLineItem
            start={new Date('2022-05-01')}
            periodColor={'#FF6666'}
        />
    )
        .toJSON();
    expect(TimeLineItemDom)
        .toMatchSnapshot();
});

test('renders TimeLineItem with  end', () => {
    const TimeLineItemDom = renderer.create(
        <TimeLineItem
            start={new Date('2022-05-01')}
            end={new Date('2022-10-01')}
            periodColor={'#FF6666'}
        />
    )
        .toJSON();
    expect(TimeLineItemDom)
        .toMatchSnapshot();
});

describe(' TimeLineItem -> getBaseRectFull ', () => {
    expect(getBaseRectFull(ScreenDevice.PC))
        .toEqual(8);
    expect(getBaseRectFull(ScreenDevice.Mobile))
        .toEqual(4);
    expect(getBaseRectFull(ScreenDevice.A4))
        .toEqual(6);
});
