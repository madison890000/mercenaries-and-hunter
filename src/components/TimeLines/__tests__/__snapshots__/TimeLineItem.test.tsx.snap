// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders TimeLineItem with  end 1`] = `
<svg
  height={20}
  width={36}
>
  <g>
    <rect
      fill="#FF6666"
      height="10"
      width={5}
      x={0}
      y="5"
    />
    <rect
      fill="#FF6666"
      height="10"
      width={5}
      x={6}
      y="5"
    />
    <rect
      fill="#FF6666"
      height="10"
      width={5}
      x={12}
      y="5"
    />
    <rect
      fill="#FF6666"
      height="10"
      width={5}
      x={18}
      y="5"
    />
    <rect
      fill="#FF6666"
      height="10"
      width={5}
      x={24}
      y="5"
    />
    <rect
      fill="#FF6666"
      height="10"
      width={5}
      x={30}
      y="5"
    />
  </g>
</svg>
`;

exports[`renders TimeLineItem with no end 1`] = `
<svg
  height={20}
  width={48}
>
  <g>
    <rect
      fill="#FF6666"
      height="10"
      width={5}
      x={0}
      y="5"
    />
    <rect
      fill="#FF6666"
      height="10"
      width={5}
      x={6}
      y="5"
    />
    <rect
      fill="#FF6666"
      height="10"
      width={5}
      x={12}
      y="5"
    />
    <rect
      fill="#FF6666"
      height="10"
      width={5}
      x={18}
      y="5"
    />
    <rect
      fill="#FF6666"
      height="10"
      width={5}
      x={24}
      y="5"
    />
    <rect
      fill="#FF6666"
      height="10"
      width={5}
      x={30}
      y="5"
    />
    <rect
      fill="#FF6666"
      height="10"
      width={5}
      x={36}
      y="5"
    />
    <rect
      fill="#FF6666"
      height="10"
      width={5}
      x={42}
      y="5"
    />
  </g>
</svg>
`;
